sdk/core/core/src/az_json_parser.c

```diff
@@ -131,9 +131,34 @@ typedef struct
   int16_t exp;
 } az_dec_number;
 
-AZ_NODISCARD static double az_json_number_to_double(az_dec_number const* p)
+// try to convert a number to double
+// Return OVERFLOW when exp goes higher than 19 (double can hold 2 ^ 64 / 2, so 10 ^ 19 would be the
+// max exp supported)
+AZ_NODISCARD az_result az_json_number_to_double(az_dec_number const* p, double* out)
 {
-  return p->value * pow(10, p->exp) * p->sign;
+  if (p->exp >= 19)
+  {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  double calculated_pow = 1;
+  if (p->exp > 0)
+  {
+    for (int i = 0; i < p->exp; i++)
+    {
+      calculated_pow *= 10;
+    }
+  }
+  else
+  {
+    for (int i = 0; i < -p->exp; i++)
+    {
+      calculated_pow /= 10;
+    }
+  }
+
+  *out = p->value * calculated_pow * p->sign;
```

Comment:
I think this might be wrong if `p->exp == 0`. If I am understanding this correctly, we end up returning just `p->value` basically for that, unchanged, when we should return 1.

https://www.geeksforgeeks.org/write-an-iterative-olog-y-function-for-powx-y/

Reply:
yeah, but the function is not only calculating `pow()`  (that's just first step of function and the reason of initializing `double calculated_pow = 1;`)

After calculating 10 ^ exp,  (where we can have 10 ^ 0 = 1)
we need to multiple that for the value to get the final outcome of converting a number to a double.

For the completed flow here, we are parsing an string like `-23.56e-3` , then, on parsing, we build the decimal_number struct describing the number and then we use this function to actually generate a double out of it.

so, here is one test:
```c
    az_json_parser state = { 0 };
    TEST_EXPECT_SUCCESS(az_json_parser_init(&state, AZ_SPAN_FROM_STR(" -23.56e-3")));
    az_json_token token;
    TEST_ASSERT(az_json_parser_parse_token(&state, &token) == AZ_OK);
    TEST_ASSERT(token.kind == AZ_JSON_TOKEN_NUMBER);
    TEST_ASSERT(token.value.number == -0.02356);
    TEST_ASSERT(az_json_parser_done(&state) == AZ_OK);
```
https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/core/core/test/main.c#L319

Reply:
> After calculating 10 ^ exp, (where we can have 10 ^ 0 = 1)
we need to multiple that for the value to get the final outcome of converting a number to a double.

Ah yes, makes sense. Thanks.

Reply:
yeah, why not..
I mean, we have O(n) for at most 19 ... v/s O(log 19),,, but, let's do it just for  fun..  
it does incremental multiplication.
I will just adapt it to support negative exp

Reply:
on second though, we need extra allocation for it..  
With our current solution, we only need to allocate the calculated_pow (in stack)
But if we updated we would required an extra double allocated to hold the incremental_multiplication (since we dont want to update the original value).
So, better to leave it as O(n) to avoid growing stack